refactor(api): Split motion control and configuration implementations

sfoster1 opened the pull request:
While there's a lot of irreconcilable differences between OT3 and OT2 hardware controllers (some of which have not been made), there's a surprising amount of similarity that before this PR was duplicated code. This PR hoists as much common functionality as it can into either shared parent classes or utility modules (depending on how much state they need from the controller).

The guiding concept here is that only the end child classes may access backends. Shared code in parent classes may do static checks of configuration or cached state that is entirely self-contained to the parent class, and may not explicitly rely on abstract methods that must be implemented by children (though some methods may be overridden by children). Data that comes from backends must be passed in to these inherited methods as an argument.

Where possible, calculation code that requires maybe one or two bits of state is split into pure functions in a helper utility.

All of this is backed up by extending the hardware control tests to automatically run on both OT3 and OT2 to prevent regressions.

This shouldn't be merged into `merge-gen3-pipettes`, that's just targeted to get a cleaner diff for reviewing; this should not be merged until
- [x] #9062 is merged

sfoster1 committed:
refactor(api): Share ExecutionManager, calibration

There are a couple little functions and internally-stored variables that
are really better off being mixins from the place that they're
originally implemented. They can now have their code shared between ot2
and ot3.

sfoster1 committed:
refactor(api): split instruments out of hcapi

The many functions that just managed instrument data, without talking to
the backend or the robot config, now are in their own provider. This
leaves (primarily) cache_instruments in the hardware controller, since
it is a) specific to the ot2 or will be eventually, and b) needs to talk
to the backend, and an overload for plunger_position for ot3.

sfoster1 committed:
refactor(api): motion control utilities

Some specific code in motion determination that handles pipette pairing
alongside single-pipette motion can be shared between ot3 and ot2 and
reduced in verbosity a little.

sfoster1 committed:
refactor(api): change instrument tests for refactor

Make the changes that will be required for the instrument tests ahead of
time (they should apply even before the next refactor) to convince us
that the refactor is regression-free.

Note: these tests appeared to be incorrectly doing assert
mock.called_with() rather than mock.assert_called_with(). Astoundingly,
fixing this didn't make any tests fail other than test_aspirate_flow_rate where it
looks like there was a simple parameter change typo in the test itself,
but we got lucky here.

sfoster1 committed:
refactor(api): aspirate check into parent method

sfoster1 committed:
refactor(api): split dispense functionality

sfoster1 committed:
refactor(api): split blowout functionality

sfoster1 committed:
refactor(api): remove unnecessary docstrings

We have the long docstrings in the mypy protocol, and having them also
in the api implementations duplicates them unnecessarily.

DerekMaggio commented on api/src/opentrons/hardware_control/instrument_handler.py:
```diff
@@ -0,0 +1,553 @@
+            instr_dict = instr.as_dict()
+            # TODO (spp, 2021-08-27): Revisit this logic. Why do we need to build
+            #  this dict newly every time? Any why only a few items are being updated?
+            for key in configs:
```
If you were to change the definition of `result` to 
`result: Dict[str, Any] = {key: instr_dict[key] for key in configs}`
Would this use a set comprehension which is faster?

sfoster1 replied:
i'm pretty hesitant to change stuff like this in this pr because it already moved. keeping this pr mainly moving or adding things and only changing functionality in non-new files keeps the diff easier to read both in the review stage and when it's merged.

DerekMaggio commented on api/src/opentrons/hardware_control/instrument_handler.py:
```diff
@@ -0,0 +1,553 @@
+        instr = self._attached_instruments[mount]
+        result: Dict[str, Any] = {}
+        if instr:
+            configs = [
```
Would it make sense to pull this out into a Constant instead of defining it every time?

sfoster1 replied:
see the response above

DerekMaggio commented on api/src/opentrons/hardware_control/instrument_handler.py:
```diff
@@ -0,0 +1,553 @@
+                "No pipette attached to {} mount".format(mount.name)
+            )
+
+        pos_dict: Dict = {
```
```
        pos_dict: Dict = {
            "top": top if top is not None else instr.config.top,
            "bottom": bottom if bottom is not None else instr.config.bottom,
            "blow_out": blow_out if blow_out is not None else instr.config.blow_out,
            "drop_tip": drop_tip if drop_tip is not None else instr.config.drop_tip,
        }
```

sfoster1 replied:
see the response above

DerekMaggio reviewed (commented):
I don't have enough context on all this to give it an approval. But overall it looks much cleaner. 
Just provided a few nitpicky things.

sfoster1 committed:
refactor(api): fixup types in hc

amitlissack commented on api/src/opentrons/hardware_control/api.py:
```diff
@@ -117,20 +113,26 @@ def __init__(
+
+    @staticmethod
+    def _check_type(inst: HardwareControlAPI) -> None:
+        """Do-nothing to provide early warning if the protocol is not satisfied.
```
This seems like an unfortunate accommodation to make for mypy. But at least you explained it well.

sfoster1 replied:
Yeah. Revisiting it I'm wondering if I can just change the order of the inheritance to alter the MRO.

sfoster1 replied:
Yeah in fact I totally can! hooray!

amitlissack replied:
Great news!

amitlissack commented on api/src/opentrons/hardware_control/execution_manager.py:
```diff
@@ -65,3 +66,46 @@ async def wait_for_is_running(self):
+
+    @classmethod
+    def wait_for_running(
+        cls: Type[SubclassInstance], decorated: DecoratedMethod
```
Nice

amitlissack reviewed (commented):
This is a brutally difficult PR to review. 

I like the approach and would approve if all tests and lint would pass.

sfoster1 committed:
refactor(api): inherit hw control api from proto

This actually works completely fine as long as the protocol is last in
the inheritance list so it's deprioritized in MRO.

sfoster1 committed:
refactor(api): centralize pickup behavior

Move pick up tip behavior at a high level to the instrument handler.
This is a particularly intense refactor because the pickup tip behavior
is both a) complicated and b) split up between multiple backend calls
which remain in the child class.

As with the other liquid handler centralizations, the way around this is
to make a fairly big frozen dataclass containing the arguments (or
things trivially convertible to the arguments) to the backend calls, and
the functionality is confirmed to be regression free by relying on tests.

sfoster1 committed:
refactor(api): centralize drop tip behavior

Hoist the drop-tip behavior of the two apis into common code, using a
similar approach to pick up tip and the other liquid handling commands.

amitlissack reviewed (approved):
Love it

Laura-Danielle commented on api/src/opentrons/hardware_control/api.py:
```diff
@@ -1040,59 +700,28 @@ async def move_rel(
-                    (z_axis, self._current_position[z_axis] + delta.z),
-                )
-            )
+        mhe = MustHomeError(
```
Nit: Honestly would rather have the `MustHomeError` repeated for repeatability purposes

Laura-Danielle reviewed (approved):
Nice refactor.
